### ISSUE TITLE
Unable to create link to other presenter, service PresenterFactory has not been set.

### ISSUE THREAD

**Issue by Lexinek** [TomasVotruba/nette-single-action-presenter#7]
When using etc `Kdyby\Console` i get this error - i found out it is because `injectPrimary()` is not called.

I was able to fix it with workaround, this is how my `createPresenter()` in `PresenterFactory` now looks like:
```php
	public function createPresenter($name)
	{
		$presenterClass = $this->getPresenterClass($name);

		$presenter = $this->container->createInstance($presenterClass);
		$this->container->callInjects($presenter); // this is necesarry to fix the issue

		return $presenter;
	}
```

Though im not sure if this is the best possible fix, i might add `is_callable()` check etc... Any ideas how to improve?

**Comment by TomasVotruba** [comment]
Using heavy packages likes Kdyby\Console with this light approach doesn't make sense.

Create own `bin/console` file and use Commands as services.

There is much better implementation like: https://github.com/contributte/console

**Comment by TomasVotruba** [comment]
@todo write an article about this, I see this  anti-pattern way too often

**Comment by Lexinek** [comment]
I have updated the code, so it makes sense to me:
```php
	public function createPresenter($name)
	{
		$presenterClass = $this->getPresenterClass($name);
		$presenter = $this->container->createInstance($presenterClass);

		if ($presenter instanceof UI\Presenter) {
			$this->container->callInjects($presenter);
		}

		return $presenter;
	}
```

Well, it is not only about console, i am using other packages that requires `Kdyby\Console` (at least in dev), maybe they could get replaced with something else?:
- kdyby/doctrine (requires kdyby/console)
- zenify/doctrine-migrations (requires kdyby/console in dev)
- symplify/symfony-event-dispatcher (requires kdyby/console in dev, anyways this one is even required in this package)

I see your point, but why would i use another console implementation when most of packages i am using are dependent on `kdyby/console`?

**Comment by TomasVotruba** [comment]
I wonder how to make it the simplest way: could you send PR with Kdyby\Console and some code to this package **to make it fail**?

I'd fix it then.

---

As for "requies in dev", this related only to your application or the package alone.

When you call `composer require symplify/symfony-event-dispatcher`, only its `require` section will be installed.

**Comment by Lexinek** [comment]
> As for "requies in dev", this related only to your application or the package alone.
Way i am thinking is like this - oh i have already kdyby/console in my dependencies, ok why not use it.

Though if there is any way how to replace it with something more simple and faster, i am in!

I will send PR with some presenters that are failing.

The problem is, if there is presenter which is descendant of `Nette\Application\UI\Presenter` it relies on calling `injectPrimary()` method which does not get called - in Kdyby/Console it is CliPresenter.
I had this issue as well with my ErrorPresenter, but figured out how to fix.

**Comment by TomasVotruba** [comment]
> Way i am thinking is like this - oh i have already kdyby/console in my dependencies, ok why not use it.

Because it creates huge and unnecessary technical dept and will hit you back in 1-2 years.

> Though if there is any way how to replace it with something more simple and faster, i am in!

I've linked Kdyby\Console replacement above https://github.com/contributte/console

> The problem is, if there is presenter which is descendant of `Nette\Application\UI\Presenter`...

Well, the goal of this package is to use independent presenters only. 
Do you thinks it would useful to support both approaches, e.g. for gradual refactoring?

**Comment by Lexinek** [comment]
> Do you thinks it would useful to support both approaches, e.g. for gradual refactoring?

Definitely!!
I understand the clean approach and i like it very much, but basically this package breaks any application that is using Kdyby/Console or has ANY presenter in their application that has `Nette\Application\UI\Presenter` as a parent, because PresenterFactory is replaced.

Solution for the situation is the code i wrote above, i can send PR as well with the failing presenters.

**Comment by TomasVotruba** [comment]
Ok, I see. Can you send PR then?

**Comment by Lexinek** [comment]
I will be happy to do that!

**Comment by Lexinek** [comment]
Fixed by #8
